src/main/transaction.c

```diff
@@ -1396,61 +1358,225 @@ LOCAL_HELPER void
+			exit(-1);
+		}
+
+		switch (stage->workload.type) {
 		case WORKLOAD_TYPE_I:
-			linear_writes_async(tdata, cdata, coord, stage, &adata_q);
+			adata->workload_cb = linear_writes_async;
+			linear_writes_async(adata);
 			break;
 		case WORKLOAD_TYPE_RU:
 		case WORKLOAD_TYPE_RR:
-			random_read_write_async(tdata, cdata, coord, stage, &adata_q);
+			adata->workload_cb = random_read_write_async;
+			random_read_write_async(adata);
 			break;
 		case WORKLOAD_TYPE_RUF:
-			random_read_write_udf_async(tdata, cdata, coord, stage, &adata_q);
+			adata->workload_cb = random_read_write_udf_async;
+			random_read_write_udf_async(adata);
 			break;
 		case WORKLOAD_TYPE_D:
-			linear_deletes_async(tdata, cdata, coord, stage, &adata_q);
+			adata->workload_cb = linear_deletes_async;
+			linear_deletes_async(adata);
 			break;
 		case WORKLOAD_TYPE_RUD:
-			random_read_write_delete_async(tdata, cdata, coord, stage, &adata_q);
+			adata->workload_cb = random_read_write_delete_async;
+			random_read_write_delete_async(adata);
 			break;
+		}
 	}
+}
 
-	// wait for all the async calls to finish
-	for (uint32_t i = 0; i < n_adatas; i++) {
-		queue_pop_wait(&adata_q);
+LOCAL_HELPER bool
+async_workload_is_done(async_data_t** worker_datas, uint32_t n_workers,
+		uint64_t tps)
+{
+	while (true) {
+		uint32_t worker_datas_write_i = 0;
+
+		for (uint32_t i = 0; i < n_workers; i++) {
+			int rv;
+			async_data_t* adata = worker_datas[i];
+
+			if (tps != 0) {
+				if ((rv = pthread_mutex_trylock(&adata->done_lock)) != 0) {
+					if (rv == EBUSY) {
+						worker_datas[worker_datas_write_i++] = adata;
+						continue;
+					}
+
+					blog_error("failed to lock mutex - %d\n", rv);
+					exit(-1);
+				}
+			}
+			else {
+				if ((rv = pthread_mutex_lock(&adata->done_lock)) != 0) {
+					blog_error("failed to lock mutex - %d\n", rv);
+					exit(-1);
+				}
+			}
+
+			if ((rv = pthread_mutex_unlock(&adata->done_lock)) != 0) {
+				blog_error("failed to unlock mutex - %d\n", rv);
+				exit(-1);
+			}
+
+			if ((rv = pthread_mutex_destroy(&adata->done_lock)) != 0) {
+				blog_error("failed to destroy mutex - %d\n", rv);
+				exit(-1);
+			}
+		}
+
+		// If any worker finished then we are done with transactions.
+		if (worker_datas_write_i != n_workers) {
+			if (worker_datas_write_i == 0) {
+				return true;
+			}
+			// Else - done but need to wait for other workers to finish.
+
+			n_workers = worker_datas_write_i;
+			continue;
+		}
+
+		break; // not done
 	}
-	queue_free(&adata_q);
 
-	// free the async_data structs
-	cf_free(adatas);
+	return false;
+}
+
+LOCAL_HELPER void
+async_workload_distribute(cf_clock delta_time, double* prior_contig_unworked,
+		uint64_t tps, tdata_t* tdata)
+{
+	double delta_sec = (double)delta_time / (1000000000) /*nsec per sec*/;
+	double contig_work = (delta_sec * tps) + *prior_contig_unworked;
+	uint64_t discreet_work = (uint64_t)contig_work;
+
+	if (discreet_work > 0) {
+		uint64_t target_per_thread = 4;
```

Comment:
Why 4?

Reply:
I was updating a counter per thread and this was a number that worked well enough for me in that model... Now that it is a global quota, this should go away.

Reply:
Thanks for the spot, removing this raised the TPS high from ~<1.5M to ~<1.6M - on my local setup.